### ISSUE TITLE
Orlando entity handling: Place entity  - converting Place (address) to entity

### ISSUE THREAD

**Issue by ilovan** [cwrc/CWRC-Writer#389]
staying at <PLACE><ADDRESS>
                            <ADDRLINE>53 St James's
                                        Square</ADDRLINE>
                            </ADDRESS>, <SETTLEMENT REG="London">London</SETTLEMENT><REGION REG="Middlesex"></REGION><GEOG REG="England"></GEOG></PLACE>

becomes:

    staying at <ADDRESS>
                            <ADDRLINE>53 St James's
                                        Square</ADDRLINE>
                            </ADDRESS>
     <PLACE annotationId="ent_1007" REF="https://www.google.ca/maps/place/53%20St%20James's%20Square,%20St.%20James's,%20London%20SW1Y,%20UK"><ADDRESS annotationId="ent_1007">, <SETTLEMENT REG="London">London</SETTLEMENT><REGION REG="Middlesex"></REGION><GEOG REG="England"></GEOG></ADDRESS></PLACE>.

As you can see, 

    <ADDRESS>
                            <ADDRLINE>53 St James's
                                        Square</ADDRLINE>
                            </ADDRESS>

gets added in front of the Place entity.

**Comment by ajmacdonald** [comment]
I fixed this issue, however it brings to light a different issue: the entity mapping has little/no awareness of the contents of the entity.
So, the result after converting to entity looks like this:
```
<PLACE annotationId="ent_195" REF="https://www.google.ca/maps/place/53%20St%20James's%20Dr,%20London%20SW17%207RN,%20UK"><ADDRESS><ADDRESS>
                        <ADDRLINE>53 St James's Square</ADDRLINE>
                        </ADDRESS>, <SETTLEMENT REG="London">London</SETTLEMENT><REGION REG="Middlesex">﻿</REGION><GEOG REG="England">﻿</GEOG></ADDRESS></PLACE>
```

Basically it's not set up to handle place entities that have multiple sub-elements.

**Comment by ajmacdonald** [comment]
A similar problem happens with citations in the Orlando schema.
They are required to have a `<BIBCITS>` parent, so when tagging a text selection `<BIBCITS><BIBCIT>` gets added. However if you're converting an already existing BIBCITS/BIBCIT, the entity mapper inserts the incorrect tags, e.g.:
```
<BIBCIT>[Insert bibliographic citation information here]</BIBCIT>
<BIBCITS annotationId="ent_1467"><BIBCIT annotationId="ent_1467">test</BIBCIT></BIBCITS>
```

**Comment by SusanBrown** [comment]
Sorry to have been so slow to respond to this @ajmacdonald 

What can we do about this? There will be other such cases. 

In the case of the PLACE, can the mapping not, for the Orlando schema, add the ref value to the place element and ignore the subelements? 

And for BIBCITS, could we allow conversion of BIBCIT (not BIBCITS) tags only? There is no use case I can think of when you would want to put a ref tag on a BIBCITS tag.

**Comment by ajmacdonald** [comment]
The problem is that converting an entity uses mostly the same code as inserting an entity, and the code has no awareness of the tags (e.g. ADDRESS) that are already inside the parent tag (PLACE). This is why when converting the place tag in the above example, there ends up being two address tags.

I'll need to write some code to analyze the contents during conversion, but have it be general enough to work in all potential cases.

**Comment by SusanBrown** [comment]
Is it a matter of invoking the validation service to see if the insertion is valid at that point? Or is it more complicated than that?
Please give this priority (obviously you are—thanks! and sorry that it slipped through the cracks on my end until now).

On Nov 25, 2016, at 11:10 AM, Andrew <notifications@github.com<mailto:notifications@github.com>> wrote:


The problem is that converting an entity uses mostly the same code as inserting an entity, and the code has no awareness of the tags (e.g. ADDRESS) that are already inside the parent tag (PLACE). This is why when converting the place tag in the above example, there ends up being two address tags.

I'll need to write some code to analyze the contents during conversion, but have it be general enough to work in all potential cases.

**Comment by ajmacdonald** [comment]
For BIBCITS, there is a validation issue because BIBCIT must be a child of BIBCITS. In the code

### Case 1: Inserting an Entity

```Some example text```

becomes

```Some <BIBCITS><BIBCIT>example</BIBCIT></BIBCITS> text```

### Case 2: Converting Entities

```Some <BIBCITS><BIBCIT>example1</BIBCIT> <BIBCIT>example2</BIBCIT></BIBCITS> text```

becomes (if converting on BIBCITS):

```Some <BIBCITS><BIBCIT><BIBCIT>example1</BIBCIT> <BIBCIT>example2</BIBCIT></BIBCIT></BIBCITS> text```

or (if converting on BIBCIT):

```Some <BIBCITS><BIBCITS><BIBCIT>example1</BIBCIT></BIBCITS> <BIBCIT>example2</BIBCIT></BIBCITS> text```

**Comment by SusanBrown** [comment]
Right, it is definitely messy.I’m sure you have better ideas about how to deal with it than I do. However, just in case is sparks a useful thought:

For the Orlando schema, I’d say something like “find the place tag around the existing text and modify it alone, ignoring any sub tags”

For bibcits, there is no use case in which a BIBCITS tag needs to be coverted, only BIBCIT tags. So when converting can you not work with the BIBCIT tags only?

On Nov 25, 2016, at 11:24 AM, Andrew <notifications@github.com<mailto:notifications@github.com>> wrote:


For BIBCITS, there is a validation issue because BIBCIT must be a child of BIBCITS. In the code

Case 1: Inserting an Entity

Some example text

becomes

Some <BIBCITS><BIBCIT>example</BIBCIT></BIBCITS> text

Case 2: Converting Entities

Some <BIBCITS><BIBCIT>example1</BIBCIT> <BIBCIT>example2</BIBCIT></BIBCITS> text

becomes (if converting on BIBCITS):

Some <BIBCITS><BIBCIT><BIBCIT>example1</BIBCIT> <BIBCIT>example2</BIBCIT></BIBCIT></BIBCITS> text

or (if converting on BIBCIT):

Some <BIBCITS><BIBCITS><BIBCIT>example1</BIBCIT></BIBCITS> <BIBCIT>example2</BIBCIT></BIBCITS> text
